File: Medium/304 - RangeSumQuery2D-Immutable/Solution.cs
// Question 304. Range Sum Query 2D - Immutable (https://leetcode-cn.com/problems/range-sum-query-2d-immutable/)

/// <summary>
/// 执行用时：188 ms, 在所有 C# 提交中击败了 72.73% 的用户
/// 内存消耗：35.2 MB, 在所有 C# 提交中击败了 68.18% 的用户
/// </summary>
public class NumMatrix
{
    private int[][] _sum;

    public NumMatrix(int[][] matrix) 
    {
        int M = matrix.Count();
        if (M != 0)
        {
            int N = matrix[0].Count();

            int[][] rowsSum = new int[M][];
            for (var i = 0; i < M; i++)
            {
                int[] rowSum = new int[N];
                for (var j = 0; j < N; j++)
                {
                    rowSum[j] = (j == 0 ? 0 : rowSum[j - 1]) + matrix[i][j];
                }
                rowsSum[i] = rowSum;
            }

            _sum = new int[M + 1][];
            var sumRow = new int[N + 1];
            Array.Fill(sumRow, 0);
            _sum[0] = sumRow;

            // System.Console.WriteLine("_sum: [");
            // System.Console.WriteLine($"[{string.Join(",", sumRow)}]");

            for (var i = 1; i <= M; i++)
            {
                sumRow = new int[N + 1];
                sumRow[0] = 0;
                
                for (var j = 1; j <= N; j++)
                {
                    sumRow[j] = _sum[i - 1][j] + rowsSum[i - 1][j - 1];
                }

                // System.Console.WriteLine($"[{string.Join(",", sumRow)}]");

                _sum[i] = sumRow;
            }

            // System.Console.WriteLine("]");
        }
        else
        {
            _sum = new int[0][];
        }
    }
    
    public int SumRegion(int row1, int col1, int row2, int col2) 
    {
        return _sum[row2 + 1][col2 + 1] - _sum[row1][col2 + 1] - _sum[row2 + 1][col1] + _sum[row1][col1];
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.SumRegion(row1,col1,row2,col2);
 */

File: Medium/304 - RangeSumQuery2D-Immutable/solution.ts
// Question 304. Range Sum Query 2D - Immutable (https://leetcode-cn.com/problems/range-sum-query-2d-immutable/)

/**
 * 执行用时：108 ms, 在所有 TypeScript 提交中击败了 83.33% 的用户
 * 内存消耗：44.1 MB, 在所有 TypeScript 提交中击败了 16.67% 的用户
 *
 * Acknowledgement: This solution is the TypeScript-version duplicate of the
 * [official solution](https://leetcode-cn.com/problems/range-sum-query-2d-immutable/solution/er-wei-qu-yu-he-jian-suo-ju-zhen-bu-ke-b-2z5n/).
 * I have came up with the identical solution, which is the [C# solution](./Solution.cs), but I just want to have a look how Leetcode Official
 * will write the solution.
 *
 * @class NumMatrix
 */
class NumMatrix {
  #sums: number[][];

  constructor(matrix: number[][]) {
    const m = matrix.length;
    if (m > 0) {
      const N: number = matrix[0].length;
      this.#sums = new Array(m + 1).fill(0).map(() => new Array(N + 1).fill(0));

      for (let i = 0; i < m; i++) {
        for (let j = 0; j < N; j++) {
          this.#sums[i + 1][j + 1] =
            this.#sums[i][j + 1] +
            this.#sums[i + 1][j] -
            this.#sums[i][j] +
            matrix[i][j];
        }
      }
    }
  }

  sumRegion(row1: number, col1: number, row2: number, col2: number): number {
    return (
      this.#sums[row2 + 1][col2 + 1] -
      this.#sums[row1][col2 + 1] -
      this.#sums[row2 + 1][col1] +
      this.#sums[row1][col1]
    );
  }
}
